src/persist-client/src/batch.rs

```diff
@@ -254,43 +288,40 @@ where
     #[instrument(level = "debug", name = "batch::finish", skip_all, fields(shard = %self.shard_id))]
     pub async fn finish(
         mut self,
-        upper: Antichain<T>,
+        registered_upper: Antichain<T>,
     ) -> Result<Batch<K, V, T, D>, InvalidUsage<T>> {
-        if PartialOrder::less_than(&upper, &self.lower) {
+        if PartialOrder::less_than(&registered_upper, &self.lower) {
             return Err(InvalidUsage::InvalidBounds {
                 lower: self.lower.clone(),
-                upper,
+                upper: registered_upper,
             });
         }
-        if upper.less_equal(&self.max_ts) {
+        // disallow updates with timestamps beyond the registered `upper` of the batch,
+        // with an exception made when `since` is beyond `upper` (as these may have been
+        // produced by compaction, which is allowed to produce batches with an arbitrary
+        // `since`)
+        if registered_upper.less_equal(&self.max_ts)
+            && !PartialOrder::less_equal(&registered_upper, &self.since)
```

Comment:
this section feels a bit icky to me. not sure if there's a simpler expression of the logic, or a better way to phrase the comment